.github/CONTRIBUTING.md

```diff
@@ -69,102 +67,142 @@ There are few ways you can play with the build:
-   1. If you suspect your environment is polluted, please use [Windows Sandbox](https://docs.microsoft.com/en-us/windows/security/threat-protection/windows-sandbox/windows-sandbox-overview) and/or a new Linux distro on WSL to verify
-1. If it still fails repeatedly, please [file an issue](https://github.com/microsoft/BotFramework-WebChat/issues/new/choose) to us.
+To debug race conditions:
+
+1. Append `location.reload()` to the test page so it will keep reloading the test page until it fail
+1. Navigate to the test page manually and wait until it failed and stop reloading
+
+General tips on race conditions or intermittent test failures:
+
+-  After sending a message, wait for responses from bot
+-  After a long message is shown, wait for scroll to complete
+-  Remove or speed up animations and media progress bars
 
 ## Static code analysis
 
-Run `npm run eslint` for static code analysis.
+Run `npm run precommit` for static code analysis.
 
-If you need to skip any ESLint errors, we recommend `eslint-disable-next-line`. Comments are required on why the rule is disabled.
+To skip any ESLint errors, we prefer `eslint-disable-next-line` instead of disabling a specific rule for the whole file. Comments are required on why the rule is disabled.
 
-# Final checks
+## Final checks
 
 There are checks that automation will not be able to capture. For example:
 
--  Hygiene
-
+-  Transparency
    -  Add a new log to `CHANGELOG.md`
-   -  Fill out the pull request form for traceability
+   -  Fill out the pull request form with details
+-  Hygiene
    -  Make sure imports, members, variables, etc, are sorted alphabetically
    -  Avoid one-off variables, prefer JavaScript shorthands, shorter and faster code
-   -  No global pollution:
-      -  no specific tab order,
-      -  minimize `z-index` usage
-         -  if z-index is ultimately required, use a [new stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context),
-      -  all polyfills must comply with ECMAScript/W3C standards
-
+   -  CSS
+      -  Remove unneeded CSS styles
+      -  Use CSS BEM and always namespace with `webchat__`, for example, `webchat__block__element--modifier`
+   -  Assume offline
+      -  All assets must be self-contained, and not loaded from any external URLs
+   -  No log to console unless it is a deprecation notes, warning, or error
+   -  No global pollution, for example:
+      -  No `taborder` with numbers other than `0` or `-1`
+      -  Minimize `z-index` usage
+         -  If `z-index` is a must, it must be in a [new stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context)
+      -  Be mindful when using CSS styles in a component with content from end-developers, CSS style may leak into the content, for example:
+         -  Set `font-family` as early as possible
```

Comment:
```suggestion
         -  Set `font-family` as early as possible
```
in the CSS? Component? DOM tree?

Reply:
Why?

Reply:
Yeah, makes sense. Just need to get the wording. 

'Set CSS as high in the DOM tree as possible, for example:
```
<section>
  <header>Header</header>
  <p>First</p>
  <p>Second</p>
  <footer>Footer</p>
</section>
```

Setting `font-family` in `<section>` produces fewer lines of code then setting `font-family` in `<header>`, `<p>`, and `<footer>`, due to the cascade effect.'


What do you think?

Reply:
LGTM, updated.